### PR TITLE
Add ``invalid-name`` check for ``TypeAlias`` names

### LINKED ISSUE #7081: TypeAlias not conforming to PascalCase to throw invalid-name
### Current problem

Python `typing.TypeAlias` are an alias to a type (as we all know).  Thus, I think `TypeAlias` should have to follow type naming conventions (e.g. PascalCase for classes, unless a primitive).

In other words, we should see `TypeAlias` really as not much different from a novel `class` in regards to naming conventions.  I think pylint is the tool to help enforce this!

### Desired solution

Please see the below code snippet

```python
# pylint: disable=missing-module-docstring

from threading import Thread  # Placeholder class
from typing import TypeAlias

ThreadAlias: TypeAlias = Thread  # Say nothing here
THREAD_ALIAS: TypeAlias = Thread  # Throw invalid-name
# C0103: Type alias "THREAD_ALIAS" doesn't conform to PascalCase naming style (invalid-name)
```

To restate, I think pylint should error on `CLASS_NAME_ALIAS` with `invalid-name`.

### Additional context

See the TypeAlias PEP: https://peps.python.org/pep-0613/#error-messaging

They use `ClassName`, not `CLASS_NAME`.

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Write a good description on what the PR does.
- [x] Add an entry to the change log describing the change in
  `doc/whatsnew/2/2.15/index.rst` (or ``doc/whatsnew/2/2.14/full.rst``
   if the change needs backporting in 2.14). If necessary you can write
   details or offer examples on how the new change is supposed to work.
- [x] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #7081.

### CODE REVIEW

**Review by jamesbraza — DISMISSED** [review]
I think you've done an excellent job @DanielNoord!  Cheers!

**Comment by DanielNoord** [comment]
@cdce8p Just checking in. Do you want to review this before we merge this? Or can we merge without a review by you.

File: doc/data/messages/i/invalid-name/details.rst
Comment (cdce8p) [inline doc/data/messages/i/invalid-name/details.rst:32]: `Callable` isn't used

File: doc/whatsnew/fragments/7801.new_check
Comment (cdce8p) [inline doc/whatsnew/fragments/7801.new_check:2]: `Callable`

File: pylint/checkers/base/name_checker/checker.py
Comment (cdce8p) [inline pylint/checkers/base/name_checker/checker.py:430]: `assign_type.value.elts[node.parent.elts.index(node)]` is used for both method calls (`_assigns_typevar`, `_assigns_typealias`). Maybe it would make sense to create a temp var for it.

Thinking one step further, the following is also reused.
```py
                            assign_type.targets[0]
                            .elts[node.parent.elts.index(node)]
                            .name
```
Would it make sense to combine both methods (`_assigns_typevar` and `_assigns_typealias`) and simply pass it `assign_type`, `node`, and a "mode" argument?

File: pylint/checkers/base/name_checker/checker.py
Comment (cdce8p) [inline pylint/checkers/base/name_checker/checker.py:467]: I don't think the `return` is necessary here.

File: tests/functional/ext/typing/typing_broken_callable.py
Comment (cdce8p) [inline tests/functional/ext/typing/typing_broken_callable.py:7]:
```suggestion
# pylint: disable=missing-docstring,unsubscriptable-object,invalid-name
```
Might be just me, but I always find it irritating to see a space between error codes. Both for pylint and mypy. Similarly in the other test files.

**Review by cdce8p — CHANGES_REQUESTED** [review]
> @cdce8p Just checking in. Do you want to review this before we merge this? Or can we merge without a review by you.

I had been away for a while working on other projects. If you want to reach me in these cases it's often best to DM me on Discord.

--
Regarding the PR, I left a few comments. Mostly left over references to `Callable` and an optimization suggestion. What we should think a little bit more about is the regex though. I checked the PR against Home Assistant and saw errors for `IPAddress` and `IPNetwork`. Those should be ok, IMO. Maybe just the `T` prefix should be disallowed, e.g. `TBadName`?

File: doc/data/messages/i/invalid-name/details.rst
Comment (DanielNoord) [reply doc/data/messages/i/invalid-name/details.rst:32]: Oops, missed that.

File: tests/functional/ext/typing/typing_broken_callable.py
Comment (DanielNoord) [reply tests/functional/ext/typing/typing_broken_callable.py:7]: Oh wow, I actually thought the preferred style was to separate them. I'm fine either way but thought that was the style we were going for.

File: pylint/checkers/base/name_checker/checker.py
Comment (DanielNoord) [reply pylint/checkers/base/name_checker/checker.py:430]: I made a new temp var.

Merging doesn't make too much sense I think. The inference is different `utils.safe_infer(node.func)` and `utils.safe_infer(node)` and we check for different `qnames`. Based on the `assign_type` we would simply follow completely different code paths.

File: tests/functional/ext/typing/typing_broken_callable.py
Comment (Pierre-Sassoulas) [reply tests/functional/ext/typing/typing_broken_callable.py]: Yeah I  try to separate them myself :smile: (like arguments in a function call)

**Comment by DanielNoord** [comment]
@cdce8p Changed the regex. Let me know what you think.

This doesn't allow `TBADname` but I don't think there is a good way to reliably distinguish between `TESTType` and `TTest`. As in, when a `T` in the first place is a prefix or part of a longer word.

**Comment by DanielNoord** [comment]
After the `primer` comment I think it's better to not worry about the `T`. `TokenInfo` such as in `black` shouldn't really be flagged. Making changes.

File: pylint/checkers/base/name_checker/checker.py
Comment (cdce8p) [inline pylint/checkers/base/name_checker/checker.py:44]:
```suggestion
    "typealias": re.compile(r"^_{0,2}(?!T[A-Z]|Type)[A-Z]+[a-z]+(?:[A-Z][a-z]+)*$"),
```
Would this be a solution?
The negative lookahead would prevent cases like `TBadName` and `TypeTodo` but still allow `TodoType`. `TodoType` and `TypeTodo` could also be added as test cases.

File: tests/functional/ext/typing/typing_broken_callable.py
Comment (cdce8p) [reply tests/functional/ext/typing/typing_broken_callable.py]: To be fair I've seen (and used) both 😄 For files I write myself, I do prefer without spaces. However if I'm just editing a random test file, I usually continue with the style that is present.

File: pylint/checkers/base/name_checker/checker.py
Comment (DanielNoord) [reply pylint/checkers/base/name_checker/checker.py:44]: Yeah, although it would still disallow `TTCircuit` as in the circuit of a [tt race](https://en.wikipedia.org/wiki/Isle_of_Man_TT).

That's really far fetched but I don't really like how this is such an arbitrary rule that is really difficult to figure out if you stumble upon it. I don't think `TTCircuit` will be used very often but there might be real uses of `TT` at the start and I think it would be very frustrating to figure out why `pylint` disallows them.

What would you think about using the current pattern as the default value and using your suggestion as our own overwrite? We could potentially even add our own overwrite in the documentation of the message?

Comment (cdce8p) [reply pylint/checkers/base/name_checker/checker.py:44]: Getting the regex 100% correct is pretty difficult. What should we do for the pandas cases then (`DatetimeNaTType`, `IntervalOrNA`, ...)?

IMO we should try to find something that works for most cases and can prevent common issues. The remaining ones should be handled with `pylint: disable[invalid-name]`. In this case I believe it's a much more prevalent error to use a `T` or `Type` prefix then something like `TTCircuit`, so I think it would be worth it.

I just opened #7322 to fix the same issue for the `TypeVar` name style as I've seen `_TRemote` and `_TCommand` in HomeAssistant lately.

Comment (DanielNoord) [reply pylint/checkers/base/name_checker/checker.py:44]: I agree that `T` and `Type` will be much more prevalent but I feel like flagging a `TTCircuit` would be considered a false positive. That's why I think we might want to be a little less strict by default and let users be more strict themselves. For me it falls within the false negatives vs. false positives preference, where I think the former is always more preferable.

Comment (cdce8p) [reply pylint/checkers/base/name_checker/checker.py:44]: > For me it falls within the false negatives vs. false positives preference, where I think the former is always more preferable.

Usually I agree here. However, I would consider `invalid-name` to be at least somewhat special. It's one of the few checks which creates false-positives in order to highlight certain errors and I would think users might already be aware of that. Just an example, in HomeAssistant about 7-8% of all `pylint: disable` comments are for `invalid-name` (53 of 697).

**Comment by Pierre-Sassoulas** [comment]
Check fail should be fixed by https://github.com/PyCQA/pylint/pull/7321

File: pylint/checkers/base/name_checker/checker.py
Comment (DanielNoord) [reply pylint/checkers/base/name_checker/checker.py:44]: Hmm doesn't that indicate that `pylint` is being too strict then? The main issue I have is that it is pretty hard to know why `TTCircuit` failed based on the `doesn't conform to predefined naming style` message. If we can find a good way to communicate that T's aren't allowed without having to open the pylint documentation I would be more than fine with disallowing them. We have some other issues about `invalid-name` and it being unclear why stuff is being flagged as well, I'd like to avoid those.

Comment (Pierre-Sassoulas) [reply pylint/checkers/base/name_checker/checker.py:44]: Also, how would that work with the new design proposed in #7305 ? Explaining better why invalid-name was triggered is a big priority.

Comment (cdce8p) [reply pylint/checkers/base/name_checker/checker.py:44]: I don't think it's that big of deal in practice. At least in my experience the `pylint: disable=invalid-name` is added quickly and everybody is happy.

I think of this check more like a pointer
```
Hey here could be an error, please check that.
But if that's what you want, it's also fine.
```

Maybe we need a new error category for formatting / style issues 🤔

Comment (Pierre-Sassoulas) [reply pylint/checkers/base/name_checker/checker.py:44]: (I was thinking of #2018  is the second most up-voted issues in the repo, but the regex is hiding the fact that the name is too short, #7305 aims to separate the two. Explaining *everything* in the regex is lower priority indeed)

Comment (Pierre-Sassoulas) [reply pylint/checkers/base/name_checker/checker.py:44]: > Hey here could be an error, please check that. But if that's what you want, it's also fine.

Isn't that "convention" ? It's ok to not follow convention.

Comment (cdce8p) [reply pylint/checkers/base/name_checker/checker.py:44]: > > Hey here could be an error, please check that. But if that's what you want, it's also fine.
> 
> Isn't that "convention" ? It's ok to not follow convention.

True 😄
In that case, we might only need to better explain that it's just a convention and nothing more.

Comment (Pierre-Sassoulas) [reply pylint/checkers/base/name_checker/checker.py:44]: The issue might be that pep8 is wildly adopted (to a point that a lot of junior dev allow more time for PEP8 than to writing code that actually bring value). But PEP8 does not say anything about typevar/typealias yet so we're mixing a very popular convention with one that is less established.

Comment (DanielNoord) [reply pylint/checkers/base/name_checker/checker.py:44]: I have updated to use @cdce8p's suggestion and hope that the additional comments in `details.rst` will prevent any issues. I think Marc is right that we should try to enforce good behaviour here rather than prevent 2/3 opened issues about this.

**Comment by jamesbraza** [comment]
Can I help forward this at all?  I hate to see all this work go stale for 4 months

**Review by Pierre-Sassoulas — DISMISSED** [review]
LGTM, thank you for reminding us @jamesbraza. We have a lot going on and some MR that are hard/long to review get lost in the flow.

**Comment by cdce8p** [comment]
> Can I help forward this at all? I hate to see all this work go stale for 4 months

I'll take another look at it soon.

**Comment by DanielNoord** [comment]
> > Can I help forward this at all? I hate to see all this work go stale for 4 months
> 
> I'll take another look at it soon.

@cdce8p Do you think you'll be able to have a look at this still?

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I would prefer to merge this even without the final review from Marc. If we get feedback on the regex we can always change it, but the underlying changes seem good to me. Are you okay with this?

**Comment by Pierre-Sassoulas** [comment]
Yeah allright, let's merge.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by DanielNoord** [comment]
Had to completely rebase as I was getting a lot of CI issues. Sorry for the loss of history, but this was the most sane way to do this.

**Comment by DanielNoord** [comment]
Also note that the recent PR broke `main`. I included a commit to fix this.

**Pull request merged**
